### PR TITLE
[SYCL][ESIMD] Add support for lsc mem access APIs

### PR DESCRIPTION
**Pull request by sndmitriev**
Signed-off-by: Sergey Dmitriev <serguei.n.dmitriev@intel.com>

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 1]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+  constexpr detail::lsc_data_order _Transposed =
+      detail::lsc_data_order::nontranspose;
+  simd<uintptr_t, N> addrs = reinterpret_cast<uintptr_t>(p);
+  addrs += convert<uintptr_t>(offsets);
+  return __esimd_lsc_load_stateless<T, L1H, L3H, _AddressScale, _ImmOffset, _DS,
+                                    _VS, _Transposed, N>(pred.data(),
+                                                         addrs.data());
+}
+
+/// Flat-address transposed gather with 1 channel.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1476]: Please replace "flat-address" to "USM pointer" here and in other places in the follow-up PR.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 2]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+      detail::lsc_data_order::transpose;
+  constexpr int N = 1;
+  simd_mask<N> pred = 1;
+  simd<uintptr_t, N> addrs = reinterpret_cast<uintptr_t>(p);
+  __esimd_lsc_store_stateless<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
+                              _Transposed, N>(pred.data(), addrs.data(),
+                                              vals.data());
+}
+
+/// 2D flat-address block load.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1898]: Some usage examples are highly desirable for this and few other lsc APIs. Please add a TODO marker in the follow-up PR.

**Review by kbobrovs — APPROVED** [review]
I think this one is good to go, given that it has been internally reviewed as well.
But please follow up with another PR addressing some documentation-related comments.

**Comment by kbobrovs** [comment]
@bader, @vladimirlaz - CUDA testing stalled. This patch does can't affect CUDA - could you please merge?

File: sycl/include/sycl/ext/intel/experimental/esimd/common.hpp [hunk 3]
```diff
@@ -176,11 +176,279 @@ enum class atomic_op : uint8_t {
...
+  fsub = 0x14,
+  load = 0x15,
+  store = 0x16,
   /// Decrement: <code>*addr = *addr - 1</code>. The only operation which
   /// returns new value of the destination rather than old.
   predec = 0xff,
 };
 
+/// The scope that lsc_fence operation should apply to
+/// Supported platforms: XeHP_SDV, DG2, PVC
```
Comment (petercad) [inline sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:189]: DG2 is the first platform with LSC:
```suggestion
/// Supported platforms: DG2, PVC
```

File: sycl/include/sycl/ext/intel/experimental/esimd/common.hpp [hunk 4]
```diff
@@ -176,11 +176,279 @@ enum class atomic_op : uint8_t {
...
+    return lsc_vector_size::n1;
+  }
+}
+
+template <typename T, lsc_data_size DS>
+constexpr lsc_data_size finalize_data_size() {
+  check_lsc_data_size<T, DS>();
+  if (DS != lsc_data_size::default_size)
+    return DS;
+  else if (sizeof(T) == 1)
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:427]: Actually U8 and U16 data size values are only supported by 2d block messages. D32U8 and D32U16 should be used instead

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 5]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+/// @tparam L1H is L1 cache hint.
+/// @tparam L3H is L3 cache hint.
+/// @tparam N is the number of channels (platform dependent).
+/// @param offsets is the zero-based offsets for SLM buffer in bytes.
+/// @param pred is predicates.
+/// @return is a vector of type T and size N * NElts
+///
+template <typename T, uint8_t NElts = 1,
+          lsc_data_size DS = lsc_data_size::default_size,
+          cache_hint L1H = cache_hint::none, cache_hint L3H = cache_hint::none,
```
Comment (petercad) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1295]: Cache hints don't apply for SLM accesses -- should these template parameters be removed?

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 6]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+  constexpr int N = 1;
+  simd_mask<N> pred = 1;
+  simd<uintptr_t, N> addrs = reinterpret_cast<uintptr_t>(p);
+  __esimd_lsc_store_stateless<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
+                              _Transposed, N>(pred.data(), addrs.data(),
+                                              vals.data());
+}
+
+/// 2D flat-address block load.
+/// Supported platforms: XeHP_SDV, DG2, PVC
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1899]: 2D block messages are supported only by PVC so far

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 7]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+///
+/// Collects elements located at specified address and returns them
+/// as a single \ref simd object.
+///
+/// @tparam T is element type.
+/// @tparam BlockWidth is the block width in number of elements.
+/// @tparam BlockHeight is the block height in number of elements.
+/// @tparam NBlocks is the number of blocks.
+/// @tparam Transposed is the transposed version or not.
+/// @tparam Transformed is apply VNNI transform or not.
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1910]: VNNI transform is only supported when data size is U8 or U16

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 8]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+/// VISA instruction: lsc_load_block2d.ugm
+///
+/// Collects elements located at specified address and returns them
+/// as a single \ref simd object.
+///
+/// @tparam T is element type.
+/// @tparam BlockWidth is the block width in number of elements.
+/// @tparam BlockHeight is the block height in number of elements.
+/// @tparam NBlocks is the number of blocks.
+/// @tparam Transposed is the transposed version or not.
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1909]: Transposed messages are only supported for U32 and U64 (with some constraints)

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 9]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+/// Supported platforms: XeHP_SDV, DG2, PVC
+/// VISA instruction: lsc_store_block2d.ugm
+///
+/// Stores elements at specified address.
+///
+/// @tparam T is element type.
+/// @tparam BlockWidth is the block width in number of elements.
+/// @tparam BlockHeight is the block height in number of elements.
+/// @tparam Transposed is the transposed version or not.
+/// @tparam Transformed is apply VNNI transform or not.
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2017]: Actually neither Transposed nor Transformed attribute supported by store messages

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 10]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+#endif // __SYCL_DEVICE_ONLY__
+}
+
+/// Accessor-based atomic.
+/// Supported platforms: XeHP_SDV, DG2, PVC
+/// VISA instruction: lsc_atomic_<OP>.ugm
+///
+/// @tparam T is element type.
+/// @tparam Op is operation type.
+/// @tparam NElts is the number of elements per address.
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2218]: Atomics only support `NElts == 1`

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 11]
```diff
@@ -1254,6 +1255,1196 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+/// @tparam L3H is L3 cache hint.
+/// @tparam N is the number of channels (platform dependent).
+/// @tparam AccessorTy is the \ref sycl::accessor type.
+/// @param acc is the SYCL accessor.
+/// @param offsets is the zero-based offsets.
+/// @param pred is predicates.
+///
+template <typename T, atomic_op Op, uint8_t NElts = 1,
+          lsc_data_size DS = lsc_data_size::default_size,
+          cache_hint L1H = cache_hint::none, cache_hint L3H = cache_hint::none,
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2184]: L1 cache hint shall be either default or uncached

**Comment by kbobrovs** [comment]
> @bader, @vladimirlaz - CUDA testing stalled. This patch does can't affect CUDA - could you please merge?

I need to take this request back - few more comments arrived.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 1 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/common.hpp (hunk 3 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/common.hpp]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 5 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp]: Removed cache hints from user-visible SLM APIs.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 6 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp]: Removed XeHP_SDV and DG2 from the list of supported platforms for 2d lsc intrinsics.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 10 above)
Comment (vmustya) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2196]: I'm okay with removing if it breaks no customer code

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 11 above)
Comment (vmustya) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2162]: > Does it apply to all atomic operations (excluding SLM)?

Yes

> Does it apply to all atomic operations (excluding SLM)? Should `L1H` parameter be removed atomic templates?

I don't really know. Probably we need just a static assert

Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp]: Removed L1 cache hint from all atomics.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 9 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp]: Removed `Transposed` and `Transformed` from thus template.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 7 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1898]: I have added static asserts to check supported data sizes in https://github.com/intel/llvm/pull/5512/commits/e9e39f7c23c2f1dc2180ca546420b02363cbda0a

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 8 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1897]: I have added static asserts to check supported data sizes when Transposed is true in https://github.com/intel/llvm/pull/5512/commits/e9e39f7c23c2f1dc2180ca546420b02363cbda0a

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 12]
```diff
@@ -1254,6 +1255,1150 @@ raw_send_store(simd<T1, n1> msgSrc0, uint32_t exDesc, uint32_t msgDesc,
...
+/// Transposed SLM gather with 1 channel.
+/// Supported platforms: DG2, PVC
+/// VISA instruction: lsc_load.slm
+///
+/// Collects elements located at slm and returns them
+/// as a single \ref simd object.
+///
+/// @tparam T is element type.
+/// @tparam NElts is the number of elements to load per address.
+/// @tparam DS is the data size.
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1317]: All the Transposed (block) messages support only `u32` and `u64` values for `DS`

Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:1318]: Added static assert to all Transposed functions in https://github.com/intel/llvm/pull/5512/commits/f922b01588bc7888c5635d0c4195daeeeeeac872

File: sycl/include/sycl/ext/intel/experimental/esimd/common.hpp (hunk 4 above)
Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:427]: I think this has been fixed in https://github.com/intel/llvm/pull/5512/commits/f922b01588bc7888c5635d0c4195daeeeeeac872

Comment (vmustya) [reply sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:427]: Probably the fixed code will not work. When `sizeof(T)` is 1 or 2, values should be also bit-casted to `uint32_t`.

Comment (sndmitriev) [reply sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:428]: Please review the updated patch. I think this problem has been fixed.

**Review by kbobrovs — APPROVED** [review]

**Comment by kbobrovs** [comment]
The precommit failures is unrelated - timeouts on two tests:
AtomicRef/sub.cpp
AtomicRef/add.cpp

Probably caused by enormous amount of warning messages.
So, I'm merging the patch.

**Pull request merged**